[Writing Tools] Support compiling ENABLE_WRITING_TOOLS on the public SDK (Part 2)

rr-codes opened the pull request:
#### c82dcaa5c84f2b04b05756afbf2e3fa7b3356392
<pre>
[Writing Tools] Support compiling ENABLE_WRITING_TOOLS on the public SDK (Part 2)
<a href="https://bugs.webkit.org/show_bug.cgi?id=291798">https://bugs.webkit.org/show_bug.cgi?id=291798</a>
<a href="https://rdar.apple.com/149620168">rdar://149620168</a>

Reviewed by Wenson Hsieh.

More work towards supporting compiling ENABLE_WRITING_TOOLS on the public SDK.

The primary change in this PR is to refactor and rename the platform agnostic type definitions of `PlatformWritingToolsBehavior`
and `PlatformWritingToolsResultOptions` into their own file because of several reasons:

* The types they abstract over are API, and so they don&apos;t even make sense in the SPI folder/file

* There is precedent for files that are platform-agnostic abstractions, such as ColorCocoa and
PlatformView/CocoaView (the fact there&apos;s two of these is weird itself but that&apos;s a different issue).

* When the SPI file becomes the fake umbrella header for the clang module that will be added in a subsequent PR,
it musn&apos;t contain any C++, which `using` declarations are.

There are several additional miscellaneous fixes in support of being able to build for the public SDK,
which will be detailed below.

* Source/WebCore/PAL/pal/spi/cocoa/WritingToolsSPI.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/cocoa/CocoaWritingToolsTypes.h: Added.

Move and rename the definitions of `PlatformWritingToolsBehavior` and `PlatformWritingToolsResultOptions`.

* Source/WebKit/Platform/spi/mac/AppKitSPI.h:
* Tools/TestWebKitAPI/Tests/TestWebKitAPI/mac/AppKitSPI.h:

Move the forward declaration of some methods to the main `AppKitSPI.h` file so that it can be referenced
from WebKit in addition to TestWebKitAPI.

* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView allowedWritingToolsResultOptions]):
(-[WKWebView writingToolsBehavior]):

Use renamed types.

(-[WKWebView intelligenceTextEffectCoordinator:rectsForProofreadingSuggestionsInRange:completion:]):

The `[NSValue valueWithRect:]` method is oddly not available on the public SDK on iOS, so replace this
with the convenience form of `createNSArray` which is actually nicer and shorter anyways, and does the proper thing.

Add includes that are now needed after removing them from the header.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration setWritingToolsBehavior:]):
(-[WKWebViewConfiguration writingToolsBehavior]):

Use renamed types.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:

Remove some unnecessary includes, and add a new necessary one.

* Source/WebKit/UIProcess/Cocoa/PlatformWritingToolsUtilities.h:
* Source/WebKit/UIProcess/Cocoa/PlatformWritingToolsUtilities.mm:
(WebKit::convertToCocoaWritingToolsBehavior):
(WebKit::convertToWebWritingToolsBehavior):
(WebKit::convertToPlatformWritingToolsBehavior): Deleted.

Rename types and functions, add needed includes.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView allowedWritingToolsResultOptions]):
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(-[WebHTMLView writingToolsBehavior]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(-[WritingToolsWKWebView initWithHTMLString:]):
(-[WritingToolsWKWebView initWithHTMLString:writingToolsBehavior:]):
(-[WritingToolsWKWebView initWithHTMLString:writingToolsBehavior:attachmentElementEnabled:]):
(-[WritingToolsWKWebView allowedWritingToolsResultOptionsForTesting]):
(-[WritingToolsWKWebView writingToolsBehaviorForTesting]):
(TEST(WritingTools, ProofreadingWithUntitledImageAttachment)):
(TEST(WritingTools, CompositionWithImageAttachmentRoundTrip)):
(TEST(WritingTools, CompositionWithNonImageAttachmentRoundTrip)):
(TEST(WritingTools, WantsInlineEditing)):
(TEST(WritingTools, WritingToolsBehaviorNonEditableWithSelection)):
(TEST(WritingTools, WritingToolsBehaviorWithNoSelection)):
(TEST(WritingTools, WritingToolsBehaviorEditableWithSelection)):
(TEST(WritingTools, AllowedInputOptionsNonEditable)):
(TEST(WritingTools, AllowedInputOptionsEditable)):
(TEST(WritingTools, AllowedInputOptionsRichText)):
(TEST(WritingTools, AllowedInputOptionsPlainText)):
(TEST(WritingTools, EphemeralSessionWithDifferingTextLengths)):
(TEST(WritingTools, EphemeralSessionWithDeeplyNestedContent)):
(TEST(WritingTools, APIWithBehaviorNone)):
(TEST(WritingTools, APIWithBehaviorDefault)):
(TEST(WritingTools, APIWithBehaviorComplete)):
(TEST(WritingTools, ShowAffordance)):
(TEST(WritingTools, ShowAffordanceForMultipleLines)):
(TEST(WritingTools, ShowPanelWithNoSelection)):
(TEST(WritingTools, ShowPanelWithCaretSelection)):
(TEST(WritingTools, ShowPanelWithRangedSelection)):
(TEST(WritingTools, DISABLED_ShowToolWithRangedSelection)):
(TEST(WritingTools, DISABLED_ShowInvalidToolWithRangedSelection)):
(TEST(WritingTools, DISABLED_ContextMenuItemsNonEditable)):
(TEST(WritingTools, DISABLED_ContextMenuItemsEditable)):
(TEST(WritingTools, DISABLED_ContextMenuItemsEditableEmpty)):
(TEST(WritingTools, AppMenuNonEditable)):
(TEST(WritingTools, AppMenuEditable)):
(TEST(WritingTools, AppMenuEditableEmpty)):
(TEST(WritingTools, PDFTextSelections)):

Rename types/methods.

Canonical link: <a href="https://commits.webkit.org/293899@main">https://commits.webkit.org/293899@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/edea3dc681e42611ff9a79853e03ded57931c53b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100234 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19883 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10181 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105364 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50817 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102275 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20188 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28357 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76311 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33370 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103241 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15440 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90535 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56671 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15260 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8534 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50185 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85168 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8616 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107723 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27347 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/20043 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/85263 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27710 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86742 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84802 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21548 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29483 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7218 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21220 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27284 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32517 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27095 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30411 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28654 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/platform/cocoa/PlatformWritingToolsTypes.h:
```diff
@@ -0,0 +1,66 @@
+
+#if PLATFORM(IOS_FAMILY)
+
+typedef UIWritingToolsBehavior PlatformWritingToolsBehavior;
```
Nit - I think `using` statements are generally preferred over `typedef` in new code.

rr-codes replied:
ah ok, will change. What's the reasoning behind this out of curiosity?

whsieh replied:
`using` is more expressive than `typedef` (e.g. works for templated types as well), and (generally) using one way to define type aliases in the codebase makes searching for type definitions a bit easier.

`typedef` does have an advantage in that it's compatible with C, but since this is a project header in WebCore, that's not a concern.

rr-codes committed:
[Writing Tools] Support compiling ENABLE_WRITING_TOOLS on the public SDK (Part 2)
https://bugs.webkit.org/show_bug.cgi?id=291798
rdar://149620168

Reviewed by Wenson Hsieh.

More work towards supporting compiling ENABLE_WRITING_TOOLS on the public SDK.

The primary change in this PR is to refactor and rename the platform agnostic type definitions of `PlatformWritingToolsBehavior`
and `PlatformWritingToolsResultOptions` into their own file because of several reasons:

* The types they abstract over are API, and so they don't even make sense in the SPI folder/file

* There is precedent for files that are platform-agnostic abstractions, such as ColorCocoa and
PlatformView/CocoaView (the fact there's two of these is weird itself but that's a different issue).

* When the SPI file becomes the fake umbrella header for the clang module that will be added in a subsequent PR,
it musn't contain any C++, which `using` declarations are.

There are several additional miscellaneous fixes in support of being able to build for the public SDK,
which will be detailed below.

* Source/WebCore/PAL/pal/spi/cocoa/WritingToolsSPI.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/cocoa/CocoaWritingToolsTypes.h: Added.

Move and rename the definitions of `PlatformWritingToolsBehavior` and `PlatformWritingToolsResultOptions`.

* Source/WebKit/Platform/spi/mac/AppKitSPI.h:
* Tools/TestWebKitAPI/Tests/TestWebKitAPI/mac/AppKitSPI.h:

Move the forward declaration of some methods to the main `AppKitSPI.h` file so that it can be referenced
from WebKit in addition to TestWebKitAPI.

* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView allowedWritingToolsResultOptions]):
(-[WKWebView writingToolsBehavior]):

Use renamed types.

(-[WKWebView intelligenceTextEffectCoordinator:rectsForProofreadingSuggestionsInRange:completion:]):

The `[NSValue valueWithRect:]` method is oddly not available on the public SDK on iOS, so replace this
with the convenience form of `createNSArray` which is actually nicer and shorter anyways, and does the proper thing.

Add includes that are now needed after removing them from the header.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration setWritingToolsBehavior:]):
(-[WKWebViewConfiguration writingToolsBehavior]):

Use renamed types.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:

Remove some unnecessary includes, and add a new necessary one.

* Source/WebKit/UIProcess/Cocoa/PlatformWritingToolsUtilities.h:
* Source/WebKit/UIProcess/Cocoa/PlatformWritingToolsUtilities.mm:
(WebKit::convertToCocoaWritingToolsBehavior):
(WebKit::convertToWebWritingToolsBehavior):
(WebKit::convertToPlatformWritingToolsBehavior): Deleted.

Rename types and functions, add needed includes.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView allowedWritingToolsResultOptions]):
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(-[WebHTMLView writingToolsBehavior]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(-[WritingToolsWKWebView initWithHTMLString:]):
(-[WritingToolsWKWebView initWithHTMLString:writingToolsBehavior:]):
(-[WritingToolsWKWebView initWithHTMLString:writingToolsBehavior:attachmentElementEnabled:]):
(-[WritingToolsWKWebView allowedWritingToolsResultOptionsForTesting]):
(-[WritingToolsWKWebView writingToolsBehaviorForTesting]):
(TEST(WritingTools, ProofreadingWithUntitledImageAttachment)):
(TEST(WritingTools, CompositionWithImageAttachmentRoundTrip)):
(TEST(WritingTools, CompositionWithNonImageAttachmentRoundTrip)):
(TEST(WritingTools, WantsInlineEditing)):
(TEST(WritingTools, WritingToolsBehaviorNonEditableWithSelection)):
(TEST(WritingTools, WritingToolsBehaviorWithNoSelection)):
(TEST(WritingTools, WritingToolsBehaviorEditableWithSelection)):
(TEST(WritingTools, AllowedInputOptionsNonEditable)):
(TEST(WritingTools, AllowedInputOptionsEditable)):
(TEST(WritingTools, AllowedInputOptionsRichText)):
(TEST(WritingTools, AllowedInputOptionsPlainText)):
(TEST(WritingTools, EphemeralSessionWithDifferingTextLengths)):
(TEST(WritingTools, EphemeralSessionWithDeeplyNestedContent)):
(TEST(WritingTools, APIWithBehaviorNone)):
(TEST(WritingTools, APIWithBehaviorDefault)):
(TEST(WritingTools, APIWithBehaviorComplete)):
(TEST(WritingTools, ShowAffordance)):
(TEST(WritingTools, ShowAffordanceForMultipleLines)):
(TEST(WritingTools, ShowPanelWithNoSelection)):
(TEST(WritingTools, ShowPanelWithCaretSelection)):
(TEST(WritingTools, ShowPanelWithRangedSelection)):
(TEST(WritingTools, DISABLED_ShowToolWithRangedSelection)):
(TEST(WritingTools, DISABLED_ShowInvalidToolWithRangedSelection)):
(TEST(WritingTools, DISABLED_ContextMenuItemsNonEditable)):
(TEST(WritingTools, DISABLED_ContextMenuItemsEditable)):
(TEST(WritingTools, DISABLED_ContextMenuItemsEditableEmpty)):
(TEST(WritingTools, AppMenuNonEditable)):
(TEST(WritingTools, AppMenuEditable)):
(TEST(WritingTools, AppMenuEditableEmpty)):
(TEST(WritingTools, PDFTextSelections)):

Rename types/methods.

Canonical link: https://commits.webkit.org/293899@main
